src/infra/core/src/utils/simple_transfer_protocol.rs

```diff
@@ -247,130 +410,41 @@ impl SimpleTransferProtocol {
-
-                let stream = match src
-                    .as_checkpoint_repo()
-                    .get_stream(&checkpoint.physical_hash)
-                    .await
-                {
-                    Ok(s) => Ok(s),
-                    Err(GetError::NotFound(e)) => Err(CorruptedSourceError {
-                        message: "Source checkpoint file is missing".to_owned(),
-                        source: Some(e.into()),
-                    }
-                    .into()),
-                    Err(GetError::Access(e)) => Err(SyncError::Access(e)),
-                    Err(GetError::Internal(e)) => Err(SyncError::Internal(e)),
-                }?;
-
-                match dst
-                    .as_checkpoint_repo()
-                    .insert_stream(
-                        stream,
-                        InsertOpts {
-                            precomputed_hash: if !trust_source_hashes {
-                                None
-                            } else {
-                                Some(&checkpoint.physical_hash)
-                            },
-                            expected_hash: Some(&checkpoint.physical_hash),
-                            // This hint is necessary only for S3 implementation that does not
-                            // currently support streaming uploads
-                            // without knowing Content-Length. We should remove it in future.
-                            size_hint: Some(checkpoint.size as usize),
-                            ..Default::default()
-                        },
-                    )
-                    .await
-                {
-                    Ok(_) => Ok(()),
-                    Err(InsertError::HashMismatch(e)) => Err(CorruptedSourceError {
-                        message: concat!(
-                            "Checkpoint file hash declared by the source didn't ",
-                            "match the computed - this may be an indication of hashing ",
-                            "algorithm mismatch or an attempted tampering",
-                        )
-                        .to_owned(),
-                        source: Some(e.into()),
-                    }
-                    .into()),
-                    Err(InsertError::Access(e)) => Err(SyncError::Access(e)),
-                    Err(InsertError::Internal(e)) => Err(SyncError::Internal(e)),
-                }?;
-
-                stats.src.checkpoints_read += 1;
-                stats.dst.checkpoints_written += 1;
-                stats.src.bytes_read += checkpoint.size as usize;
-                stats.dst.bytes_written += checkpoint.size as usize;
-                listener.on_status(SyncStage::TransferData, &stats);
-            }
-        }
+        stream::iter(block_download_tasks)
+            .for_each_concurrent(
+                SimpleProtocolTransferOptions::default().max_parallel_transfers,
+                |future| async move { future.await.unwrap() },
```

Comment:
unwrap will cause the whole process to panic if a single file download fails, let's:
- add a test that highlights this
- find a way to abort transfer upon first error
- surface the error in result